components/merkledb/src/proof_list_index/mod.rs

```diff
@@ -823,3 +824,83 @@ where
         self.base_iter.next().map(|(_, v)| v)
     }
 }
+
+#[cfg(feature = "with-protobuf")]
+#[doc(hidden)]
+pub mod proto {
+    pub use crate::proto::{self, *};
+    use failure::Error;
+    use protobuf::RepeatedField;
+    use std::borrow::Cow;
+
+    use super::{HashedEntry, ListProof, ProofListKey};
+    use crate::{proof_list_index::MAX_INDEX, BinaryValue};
+    use exonum_proto::ProtobufConvert;
+
+    impl ProtobufConvert for ProofListKey {
+        type ProtoStruct = proto::ProofListKey;
+
+        fn to_pb(&self) -> Self::ProtoStruct {
+            let mut key = proto::ProofListKey::new();
+            key.set_index(self.index());
+            key.set_height(self.height().into());
+            key
+        }
+
+        fn from_pb(pb: Self::ProtoStruct) -> Result<Self, Error> {
+            let index = pb.get_index();
+            let height = pb.get_height();
+
+            // ProtobufConvert is implemented manually to add this checks.
+            ensure!(index <= MAX_INDEX, "index is out of range");
+            ensure!(height <= 58, "height is out of range");
+
+            Ok(ProofListKey::new(index as u8, height.into()))
```

Comment:
Why are we checking that index is less than `2_u64.pow(56) - 1` to later cast it to u8?

Reply:
It was a mistake, I've changed the test to cover this case.